mex/backend/graph/hydrate.py

```diff
@@ -117,36 +119,92 @@ def hydrate(flat: FlatDict, model: type[BaseModel]) -> NestedDict:  # noqa: C901
-        model_at_depth = model
-        target = nested
-        for depth, key in enumerate(key_path, start=1):
-            if depth < key_path_length:
-                if not issubclass(model_at_depth, BaseModel):
-                    raise TypeError("cannot hydrate paths with non base models")
-                if key in model_at_depth._get_list_field_names():
-                    if not value_is_list:
-                        raise TypeError("cannot hydrate non-list to list")
-                    if isinstance(target, list):
-                        raise TypeError("cannot handle multiple list branches")
-                    target = target.setdefault(key, [{} for _ in range(value_count)])  # type: ignore
-                elif isinstance(target, list):
-                    if len(target) != value_count:  # pragma: no cover
-                        raise RuntimeError("branch count must match our values")
-                    target = [t.setdefault(key, {}) for t in target]
-                else:
-                    target = target.setdefault(key, {})  # type: ignore
-            elif isinstance(target, list):
-                for t, v in zip(target, value):
-                    t[key] = hydrate_value(v)
-            else:
-                target[key] = hydrate_value(value)  # type: ignore
+        empty_leaf_value = _initialize_branch_with_missing_expected_types(
+            branch_keys, model, nested, value_count, value_is_list
+        )
+
+        _set_leaf_values(empty_leaf_value, leaf_key, value)
+
+    return nested
+
+
+def _initialize_branch_with_missing_expected_types(
+    branch_keys: list[str],
+    model: type[BaseModel],
+    nested: NestedDict,
+    value_count: int,
+    value_is_list: bool,
+) -> NestedDict | list[NestedDict]:
+    model_at_depth = model
+    nested_value_of_current_branch_key: NestedDict | list[NestedDict] = nested
+    for key_id, branch_key in enumerate(branch_keys):
+        nested_value_of_parent_branch_key = nested_value_of_current_branch_key
+        nested_value_of_current_branch_key = _set_branch_node_default(
+            nested_value_of_parent_branch_key,
+            branch_key,
+            model_at_depth,
+            value_count,
+            value_is_list,
+        )
+        if len(branch_keys) - key_id > 1:  # if for loop has iterations left
             try:
-                model_at_depth = model_at_depth.__fields__[key].type_
+                model_at_depth = _get_base_model_from_field(
+                    model_at_depth.model_fields[branch_key]
+                )
             except KeyError:
                 raise TypeError("flat dict does not align with target model")
-
-    return nested
+    return nested_value_of_current_branch_key
+
+
+def _get_base_model_from_field(field: FieldInfo) -> type[BaseModel]:
+    if args := get_args(field.annotation):
+        args_wo_none = [arg for arg in args if arg is not type(None)]
+        if args_count := len(args_wo_none) != 1:
```

Comment:
```suggestion
        if (args_count := len(args_wo_none)) != 1:
```

Reply:
added in 964bdfe22fe8646eeaa63f1e47b983c5e4f11d28